File: benchmarks/word-count/wordcount_simple.mjs
import fs from "node:fs";

const input = fs.readFileSync(0, "utf8");
const words = input.match(/[A-Za-z]+/g) ?? [];
const counts = new Map();

for (const raw of words) {
  const w = raw.toLowerCase();
  counts.set(w, (counts.get(w) ?? 0) + 1);
}

const entries = [...counts.entries()].map(([word, count]) => ({ word, count }));
entries.sort((a, b) => (b.count - a.count) || (a.word < b.word ? -1 : a.word > b.word ? 1 : 0));

const topN = Math.min(entries.length, 10);
let out = "";
for (let i = 0; i < topN; i += 1) {
  out += `${entries[i].count}\t${entries[i].word}\n`;
}
out += "--\n";
out += `total\t${words.length}\n`;
out += `unique\t${entries.length}\n`;
process.stdout.write(out);

File: benchmarks/word-count/wordcount_simple.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"regexp"
	"sort"
	"strings"
)

var wordRE = regexp.MustCompile(`[A-Za-z]+`)

type entry struct {
	Word  string
	Count int
}

func main() {
	in, err := io.ReadAll(os.Stdin)
	if err != nil {
		fmt.Fprintf(os.Stderr, "read stdin: %v\n", err)
		os.Exit(1)
	}

	counts := map[string]int{}
	total := 0
	for _, raw := range wordRE.FindAll(in, -1) {
		w := strings.ToLower(string(raw))
		counts[w]++
		total++
	}

	out := format(counts, total)
	if _, err := os.Stdout.Write(out); err != nil {
		fmt.Fprintf(os.Stderr, "write stdout: %v\n", err)
		os.Exit(1)
	}
}

func format(counts map[string]int, total int) []byte {
	entries := make([]entry, 0, len(counts))
	for w, c := range counts {
		entries = append(entries, entry{Word: w, Count: c})
	}
	sort.Slice(entries, func(i, j int) bool {
		if entries[i].Count != entries[j].Count {
			return entries[i].Count > entries[j].Count
		}
		return entries[i].Word < entries[j].Word
	})

	var b bytes.Buffer
	n := len(entries)
	if n > 10 {
		n = 10
	}
	for i := 0; i < n; i++ {
		fmt.Fprintf(&b, "%d\t%s\n", entries[i].Count, entries[i].Word)
	}
	fmt.Fprint(&b, "--\n")
	fmt.Fprintf(&b, "total\t%d\n", total)
	fmt.Fprintf(&b, "unique\t%d\n", len(entries))
	return b.Bytes()
}
